### PR TITLE
[Leo]: Update browser icons to use Leo Icons

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/30186

@aguscruiz has created a bunch of great icons and we want to update the browser to use his Leo Icons. These should be automatically updated when Leo is updated so he's able to update the icons with minimal effort (an update in Figma will eventually propagate all the way through to Brave Core).

### CODE REVIEW

File: patches/components-vector_icons-aggregate_vector_icons.py.patch [hunk 1]
```diff
@@ -1,8 +1,28 @@
...
++def get_overrides():
++  brave_root = f'{Path(__file__).parent}/../../brave'
++  overrides_file_path =  f'{brave_root}/vector_icons/leo_overrides.json'
++
++  leo_overrides = {}
++  data = json.load(open(overrides_file_path))
++
++  for icon, leo_override in data.items():
++    leo_overrides[icon] = f'../../brave/node_modules/@brave/leo/icons-skia/{leo_override}.icon'
++  return leo_overrides
```
Comment (mkarolin) [inline patches/components-vector_icons-aggregate_vector_icons.py.patch:21]: For python patching we typically externalize stuff like this into a separate python file (under brave/scripts) and then try to minimize the patch using an import. For example, [script](https://github.com/brave/brave-core/blob/master/script/signing_helper.py) and [patch](https://github.com/brave/brave-core/blob/master/patches/chrome-installer-mac-signing-config_factory.py.patch)

File: patches/components-vector_icons-aggregate_vector_icons.py.patch [hunk 2]
```diff
@@ -11,28 +31,51 @@ index 3e5a94a7f053ffc66f1a820073265cbfe1332b39..0175ec714b441df5e8fa58e9391e2de6
...
  
 +    # Check for alternative path
 +    alt_icon_path = os.path.join(alt_working_directory, os.path.basename(icon_path))
-+    if (os.path.exists(alt_icon_path)):
++    
++    if icon_name in leo_overrides:
++      icon_path = leo_overrides[icon_name]
++    elif os.path.exists(alt_icon_path):
 +      icon_path = alt_icon_path
 +
```
Comment (mkarolin) [inline patches/components-vector_icons-aggregate_vector_icons.py.patch:58]: I know you are just updating this, but it would be ideal to externalize the bulk of this as well.

File: patches/components-vector_icons-aggregate_vector_icons.py.patch [hunk 3]
```diff
@@ -1,8 +1,16 @@
...
-index 3e5a94a7f053ffc66f1a820073265cbfe1332b39..0175ec714b441df5e8fa58e9391e2de6bbd67755 100644
+index 3e5a94a7f053f..2ccc03b6c819e 100644
 --- a/components/vector_icons/aggregate_vector_icons.py
 +++ b/components/vector_icons/aggregate_vector_icons.py
-@@ -102,7 +102,7 @@ def ExtractIconReps(icon_file_name):
+@@ -10,6 +10,7 @@ import re
+ import shlex
+ import sys
+ import textwrap
++from import_inline import inline_file_from_src; inline_file_from_src('brave/chromium_src/components/vector_icons/aggregate_vector_icons.py', globals(), locals())
```
Comment (goodov) [inline patches/components-vector_icons-aggregate_vector_icons.py.patch:9]: pls put this just before `def main():`. It should work the same, but this way it's possible to override everything declared in the file.

File: vector_icons/leo_overrides.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  "download_in_progress": "leo_download",
+  "download_toolbar_button": "leo_download",
+  "battery_saver": "leo_energy_saver",
+  "zoom_plus": "leo_search_zoom_in",
+  "zoom_minus": "leo_search_zoom_out",
+  "not_secure_warning": "leo_warning_triangle_filled",
+  "http": "leo_warning_circle_filled",
+  "find_in_page": "leo_window_search",
+  "media_toolbar_button": "leo_music_tones",
+  "install_desktop": "leo_pwa_install"
```
Comment (goodov) [inline vector_icons/leo_overrides.json:26]: nit: this list should be sorted.

File: patches/components-vector_icons-vector_icons.gni.patch [hunk 5]
```diff
@@ -1,15 +1,15 @@
...
-@@ -42,6 +42,7 @@ template("aggregate_vector_icons") {
-       "vector_icons.cc.template",
-       "vector_icons.h.template",
-     ]
-+    alt_icons_dir = "//brave/vector_icons/" + rebase_path(invoker.icon_directory, "//") alt_icons_dir_rel = rebase_path(alt_icons_dir, root_build_dir)
+@@ -45,6 +45,7 @@ template("aggregate_vector_icons") {
      inputs =
          rebase_path(templates + invoker.sources, ".", invoker.icon_directory) +
          [ "//components/vector_icons/aggregate_vector_icons.py" ]
++    alt_icons_dir = "//brave/vector_icons/" + rebase_path(invoker.icon_directory, "//") alt_icons_dir_rel = rebase_path(alt_icons_dir, root_build_dir) inputs += ["//brave/vector_icons/leo_overrides.json"]
```
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply patches/components-vector_icons-vector_icons.gni.patch:9]: Updated to use absolute paths, as I thought it was probably easiest.

File: chromium_src/components/vector_icons/aggregate_vector_icons.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    """Determines where to read the icon from. Options are:
+    1. The Leo `icons-skia` folder, mapping `icon_name` to a Leo icon, as per
+       the `leo_overrides.json` file.
+    2. The Brave equivalent of the the Chromium icon path.
+    3. The Chromium Path (this is the default, if no override is specified).
+
+  Args:
+      icon_path: The path to the chromium icon
+      alt_working_directory: The Brave overrides folder for this icon.
+  """
```
Comment (goodov) [inline chromium_src/components/vector_icons/aggregate_vector_icons.py:50]: incorrect format

File: components/skus/browser/rs/wasm/src/lib.rs [hunk 7]
```diff
@@ -78,10 +78,7 @@ pub fn initialize(
             .with_writer(MakeConsoleWriter);
         let perf_layer = performance_layer().with_details_from_fields(Pretty::default());
 
-        tracing_subscriber::registry()
-            .with(fmt_layer)
-            .with(perf_layer)
-            .init();
+        tracing_subscriber::registry().with(fmt_layer).with(perf_layer).init();
```
Comment (goodov) [inline components/skus/browser/rs/wasm/src/lib.rs:81]: pls revert this unrelated change. You might need to do `git fetch origin master` locally.

File: vector_icons/leo_overrides.json [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  "//chrome/app/vector_icons/zoom_plus.icon": "leo_search_zoom_in",
+  "//chrome/app/vector_icons/zoom_minus.icon": "leo_search_zoom_out",
+  "//chrome/app/vector_icons/navigate_stop.icon": "leo_close",
+  "//chrome/app/vector_icons/navigate_stop_windows.icon": "leo_close",
+  "//chrome/app/vector_icons/navigate_home.icon": "leo_browser_home",
+  "//chrome/app/vector_icons/media_toolbar_button.icon": "leo_music_tones",
+  "//chrome/app/vector_icons/download_toolbar_button.icon": "leo_download",
+  "//chrome/app/vector_icons/download_in_progress.icon": "leo_download",
+  "//chrome/app/vector_icons/browser_tools.icon": "leo_hamburger_menu",
+  "//chrome/app/vector_icons/battery_saver.icon": "leo_energy_saver"
```
Comment (goodov) [inline vector_icons/leo_overrides.json:26]: sort please?

**Review by goodov — APPROVED** [review]

File: components/skus/browser/rs/wasm/src/lib.rs (hunk 7 above)
Comment (fallaciousreasoning) [reply components/skus/browser/rs/wasm/src/lib.rs:81]: Nice catch - didn't notice when I formatted it!

File: vector_icons/leo_overrides.json (hunk 8 above)
Comment (fallaciousreasoning) [reply vector_icons/leo_overrides.json:26]: Whoops, accidentally sorted the wrong way in my IDE :facepalm:
